File: HW8/GTUSet.java
import java.lang.reflect.Array;

/**
 * @author zafer
 *
 * @param <T>
 */
public class GTUSet<T> extends GTUContainer<T> {

	GTUIterator<T> i;
	public T []data;
	public T []data2;
	int maxsize=50;
	int current =0,counter=0;
	int flag=1;

	/**
	 * Returns max size
	 * @return
	 */
	public int getMaxsize() {
		return maxsize;
	}
	/**
	 * Returns current
	 * @return
	 */
	public int getCurrent() {
		return current;
	}
	/**
	 * Constructor
	 */
	public GTUSet (){
		data= (T[])new Object[maxsize];
	}
	/**
	 *If set is empty,return true
	 */
	public boolean empty(){
		if(current==0)
		return true;
		return false;
	}
	/**
	 * Returns size
	 * @return size
	 */
	public int size() {
		return current;
	}
	/**Returns max size
	 *@return max size
	 */
	public int maxSize() {
		return maxsize;
	}
	/**
	 *resets the set
	 */
	public void clear() {
		maxsize=50;
		current=0;
		counter=0;
		data=(T[])new Object[maxsize];
	}
	/**
	 * Returns numth index of set
	 * @param num
	 * @return
	 */
	public T get(int num) { 
		return data[num];
	}
	/**
	 *	Inserts num1 if not same element in set
	 *@param num1
	 *
	 *
	 */
	public void insert(T num1) {
		
		if(current==maxsize) {
			maxsize*=2;
			 T[] data2 = (T[]) Array.newInstance(data[0].getClass(), data.length * 2);
			 System.arraycopy(data, 0, data2, 0, data.length);
			 data=data2;
		}
		
		for(int i=0;i<current;i++) {
			if(num1==data[i]) {
				flag=0;
			}
		}
		if(flag==1) {
			data[current]=num1;
			current++;
			
		}
			
	}
	/**
	 *Deletes if num2 exists
	 *@param num2 
	 */
	public void erase(T num2) {
		for(int i=0;i<current;i++) {
			if(num2==data[i]) {
				counter=i;
				break;
			}
		}
		for(int j=counter;j<current;j++) {
			data[j]=data[j+1];
		}
	}
	/**
	 * It takes abstract class obj and creates iterator
	 *@param obj abstract class object
	 *@return iterator
	 */
	public GTUIterator<T> Iterator(GTUContainer<T> obj){
		i=new GTUIterator<T>(obj); 
		return i;
	}
}

File: HW6/gtuset.cpp
#include "gtuset.h"

namespace zaff{
template<class T>
GTUSet<T>::GTUSet():head(NULL),now(NULL){

}
template<class T>
GTUSet<T>::GTUSet(const GTUSet<T>& set){
	if(set.empty()){
		head=now=NULL;
	}
}
template<class T>
GTUSet<T>::~GTUSet(){
	head=NULL;
}
template<class T>
GTUSet<T>::GTUSet(int x){
	int i=1;
	shared_ptr<Node<T>> temp=shared_ptr <Node<T>> (new Node<T> (0,NULL));
	now=head=temp;
	while(i<x){
	temp=shared_ptr <Node<T>> (new Node<T> (0,NULL));
	now->setNext(temp);
	now=now->getNext();

	i++;

	}
}
template<class T>
bool GTUSet<T>::empty()const{
	if (head==NULL)
	{
		return true;
	}
	else{
		return false;
	}
}

template<class T>
GTUIterator<T> GTUSet<T>::begin()const{
	return Iterator(head);
}


template<class T>
void GTUSet<T>::clear(){
	shared_ptr <Node<T>> temp;

while(temp!=head){

		temp=head;
		while(temp->getNext()->getNext()!=NULL){
			temp=temp->getNext();
		}
		temp->setNext(NULL);
	

	}
	head=NULL;
}

template<class T>
GTUIterator<T> GTUSet<T>::end()const{
	return Iterator(now->getNext());
}

template<class T>
T GTUSet<T>::operator[](int x){
	int i=0;
	shared_ptr <Node<T>> temp;
	temp=head;
	while(i<x){
		temp=temp->getNext();
		i++;
	}
	return temp->getData();

}
template<class T>
int GTUSet<T>::maxsize(){
	int i=1;
	shared_ptr <Node<T>> temp;
	temp=head;
	while(temp->getNext()!=NULL){
		temp=temp->getNext();
		i++;
	}

	return i;	
}
template<class T>
void GTUSet<T>::insert(T data){
	int flag=0;
	if (empty())
	{
		now=shared_ptr <Node<T>> (new Node<T> (data,NULL));
		head=now;
	}
	else{
		shared_ptr <Node<T>> temp;
		shared_ptr <Node<T>> temp2;
		temp=head;
		while(temp!=NULL){	
			if (data == temp->getData())
			{	
				flag=1;
				break;
			}
				temp=temp->getNext();
		  }	 
		if (flag==0)
		{
			temp2=shared_ptr <Node<T>> (new Node<T> (data,NULL));
			now->setNext(temp2);
			now=now->getNext();
		}
	}

}

template<class T>
void GTUSet<T>::erase(T data){
	int i=0;
	shared_ptr <Node<T>> temp;
	shared_ptr <Node<T>> temp2;
	temp=head;
	while(temp->getNext()->getData()!=data){
		temp=temp->getNext();

	}
	temp2=temp->getNext()->getNext();
	temp->setNext(temp2);
}
template<class T>
int GTUSet<T>::size(){
	int i=1;
	shared_ptr <Node<T>> temp;
	temp=head;
	while(temp->getNext()!=NULL){
		temp=temp->getNext();
		i++;
	}

	return i;
}
}